src/Access/LDAPAccessStorage.cpp

```diff
@@ -312,7 +312,7 @@ void LDAPAccessStorage::updateAssignedRolesNoLock(const UUID & id, const String
         return entity_;
     };
 
-    memory_storage.update(id, update_func);
+    memory_storage.updateNoLock(id, update_func, true);
```

Comment:
The problem occurs when we lock mutex of `LDAPAccessStorage` in `authenticateImpl` and then call `update` from `updateAssignedRolesNoLock` using the same mutex - since `AccessControl` contains the same instance of `LDAPAccessStorage`. There are different solutions possible, but I guess this one is simplest and reliable.
Seems there's no danger of modifying data from different thread here since this call always locked in `LDAPAccessStorage::authenticateImpl` and `updateNoLock` can be called from single thread only.
Please correct me if I'm wrong.

Reply:
Yes, seems this is the same scenario. It was the first solution for me but I decided not to use it because using recursive mutex is not the safest idea. Well anyway it's your choice. Both solutions resolve the issue.